csrc/multidevice/pipeline.cpp

```diff
@@ -0,0 +1,307 @@
+  // stores the Vals that will be I/O of each stage
+  std::unordered_map<const PipelineStageDescriptor*, ValSet> StageInputsDesc;
+  std::unordered_map<const PipelineStageDescriptor*, ValSet> StageOutputsDesc;
+  // Store the stages that consumes a given Val produced at another stage
+  std::unordered_map<Val*, std::vector<const PipelineStageDescriptor*>>
+      ValConsumerStageDesc;
+  // maps Vals of the original fusion to their corresponding PipelineVals
+  // which are inputs/outputs of a Stage. We need to differentiate those
+  // two cases for when a stage contains only one Val, from which we create
+  // two PipelineVal
+  std::unordered_map<Val*, Val*> valToPipelineValInputOfStage;
+  std::unordered_map<Val*, Val*> valToPipelineValOutputOfStage;
+
+  // Check (partially) that the pipeline is valid and satisfies the assumptions
+  // described in pipeline.h
+  void validate() {
+    std::unordered_set<Val*> tvInStages;
+    // Check that each Val is a TensorView and that it belongs to at most one
+    // stage
+    for (auto& stage_desc : pipeline_->descriptor().stageDescriptors) {
+      for (auto& val : stage_desc->vals()) {
+        if (!val->isA<TensorView>()) {
+          TORCH_INTERNAL_ASSERT(
+              0,
+              "the Val " + val->toString() +
+                  " must be a TensorView to belong to a Pipeline");
+        } else if (tvInStages.count(val)) {
+          TORCH_INTERNAL_ASSERT(
+              0,
+              "the Val " + val->toString() + " belongs to more than one stage");
+        } else {
+          tvInStages.insert(val);
+        }
+      }
+    }
+    // Check that each TensorView belongs to at least one stage
+    for (auto& val : pipeline_->originalFusion()->vals()) {
+      TORCH_INTERNAL_ASSERT(
+          (!val->isA<TensorView>()) || tvInStages.count(val),
+          "the Val " + val->toString() + " must be added to a stage");
+    }
+  }
+
+  // returns whether a Val is an input of the originalFusion
+  bool isGlobalInput(Val* val) const {
+    return std::count(
+        pipeline_->originalFusion()->inputs().begin(),
+        pipeline_->originalFusion()->inputs().end(),
+        val);
+  }
+
+  // returns whether a Val is an output of the originalFusion
+  bool isGlobalOutput(Val* val) const {
+    return std::count(
+        pipeline_->originalFusion()->outputs().begin(),
+        pipeline_->originalFusion()->outputs().end(),
+        val);
+  }
+
+  // populates StageInputsDesc, StageOutputsDesc, ValConsumerStageDesc
+  void fillInfo() {
+    for (auto& stage_desc : pipeline_->descriptor().stageDescriptors) {
+      for (auto& val : stage_desc->vals()) {
+        // Add global inputs of the original fusion
+        if (isGlobalInput(val)) {
+          StageInputsDesc[stage_desc].pushBack(val);
+        }
+        // Add global outputs of the original fusion
+        if (isGlobalOutput(val)) {
+          StageOutputsDesc[stage_desc].pushBack(val);
+        }
```

Comment:
Does this mean fusion inputs and outputs are considered also as inputs and outputs of all pipeline stages?

Reply:
no, it means that if `tv0` is a global input of the original fusion and belongs to `stage0`, then it will be an input of `stage0`